lib/pbio/include/pbio/os.h

```diff
@@ -0,0 +1,217 @@
+ * PB_LC_SET() is done within another switch() statement!
+ */
+
+#define PB_LC_FALLTHROUGH __attribute__((fallthrough));
+#define PB_LC_INIT(state) *state = 0;
+#define PB_LC_RESUME(state) switch (*state) { case 0:
+#define PB_LC_SET(state) *state = __LINE__; PB_LC_FALLTHROUGH; case __LINE__:
+#define PB_LC_END() }
+
+/**
+ * Protothread state. Effectively the line number in the current file where it
+ * yields so it can jump there to resume later.
+ */
+typedef uint32_t pbio_os_state_t;
+
+typedef pbio_error_t (*pbio_os_process_func_t)(pbio_os_state_t *state);
+
+typedef struct _pbio_os_process_t pbio_os_process_t;
+
+/**
+ * A process.
+ */
+struct _pbio_os_process_t {
+    /**
+     * Pointer to the next process in the list.
+     */
+    pbio_os_process_t *next;
+    /**
+     * State of the protothread.
+     */
+    pbio_os_state_t state;
+    /**
+     * The protothread.
+     */
+    pbio_os_process_func_t func;
+    /**
+     * Most recent result of running one iteration of the protothread.
+     */
+    pbio_error_t err;
+};
+
+/**
+ * Initialize a protothread.
+ *
+ * Initializes a protothread. Initialization must be done prior to
+ * starting to execute the protothread.
+ *
+ * @param [in]  state    Protothread state.
+ */
+#define ASYNC_INIT(state)   PB_LC_INIT(state)
+
+/**
+ * Declare the start of a protothread inside the C function
+ * implementing the protothread.
+ *
+ * This macro is used to declare the starting point of a
+ * protothread. It should be placed at the start of the function in
+ * which the protothread runs. All C statements above the ASYNC_BEGIN()
+ * invocation will be executed each time the protothread is scheduled.
+ *
+ * @param [in]  state    Protothread state.
+ */
+#define ASYNC_BEGIN(state) { PB_LC_RESUME(state)
+
+/**
+ * Declare the end of a protothread and returns with given code.
+ *
+ * This macro is used for declaring that a protothread ends. It must
+ * always be used together with a matching ASYNC_BEGIN() macro.
+ *
+ * NB: In contrast to Contiki, this does not call ASYNC_INIT() before exiting.
+ *
+ * @param [in]  err    Error code to return.
+ */
+#define ASYNC_END(err) PB_LC_END(); return err; }
+
+/**
+ * Yields the protothread while the specified condition is true.
+ *
+ * @param [in]  state     Protothread state.
+ * @param [in]  condition The condition.
+ */
+#define AWAIT_WHILE(state, condition)            \
+    do {                                         \
+        PB_LC_SET(state);                        \
+        if (condition) {                         \
+            return PBIO_ERROR_AGAIN;             \
+        }                                        \
+    } while (0)
+
+
+#define AWAIT(state, child, statement)          \
+    do {                                        \
+        ASYNC_INIT((child));                    \
+        PB_LC_SET(state);                       \
+        if ((statement) == PBIO_ERROR_AGAIN) {  \
```

Comment:
I guess this means that `statement` can never return anything besides `PBIO_ERROR_AGAIN` or `PBIO_SUCCESS`?  Other error will be ignored, which isn't ideal. Could fix this by getting rid of `do`/`while` and introduce a local `err` variable so that this can be an expression statement. Then we could write `err = AWAIT(...);` and actually handle errors.

Reply:
I'll respond to this one on main thread.